rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,370 @@
+
+- If the original data column of an Index Function is used, transform the predicate to a new predicate based on the
+  Index Function name if possible, e.g., `timestamp between 1666767600 and 1666897200`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the expression of an Index Function is used, replace the expression to the Index Function name,
+  e.g., `DATE_FORMAT(timestamp, '%Y-%m-%d') between '2022-10-26' and '2022-10-27'`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the Index Function name is used in the predicate, e.g., `datestr between '2022-10-26' and '2022-10-27'`, no action
+  is needed here.
+
+2. For the Index Function name appeared in the predicate, look up the corresponding index metadata from metadata table
+   and prune files that are not needed for scanning.
+
+For each query engine, we need to intercept the engine-specific representation of the predicates, and derive the
+relevant predicate based on the function. We consider integrating Spark and Presto first in the implementation.
+
+#### Index-Function-Based Ordering on Write Path
+
+To unlock efficient file pruning, we plan to support Index-Function-based ordering during write operations. Similar to
+clustering, the Index functions for ordering are specified by new write config `hoodie.write.sort.columns`.
+
+When writing records to data files, the write client does the following:
+
+- For new inserts, sort the data based on the specified Index functions before writing to data files. This reuses
+  existing flow of bulk insert and clustering
+    - Each data file contains one or a small subset of values corresponding to the Index Functions.
+    - Different sort modes can be provided like bulk insert and clustering.
+
+- For updates, no change in logic is needed. The record with the same key should go into the same file group.
+- For small file handing, we need to take into account the Index functions for the workload profile, so that the new
+  inserts can be combined into the data file without expanding the minimum and maximum values of the Index Function when
+  the data file is small.
+
+## Related Concepts
+
+In this section, we compare Hudi's Index Function to related concepts and solutions on pruning files efficiently and
+improving query performance.
+
+### Hidden Partitioning
+
+[Hidden Partitioning](https://iceberg.apache.org/docs/latest/partitioning/#icebergs-hidden-partitioning), introduced by
+Apache Iceberg, maintains the relationship between the original column and transformed partition values, and hides the
+partition columns and physical partitions from catalogs and users, using manifests containing the partition values of
+each file to prune files and speed up queries. By using Hidden Partitioning, Iceberg can evolve partition schemes, i.e.,
+partition evolution, without rewriting data.
+
+Hudi's Index Function achieves the same goal of pruning files based on the column stats, with the support on any data
```

Comment:
We need to clarify this statement -- partition-pruning is relying on the data being partitioned for its efficiency. Since in our case data is unlikely to be partitioned by our derived column, we won't be able to get to the same level of efficiency of the pruning.